File: src/main/year2019/Day04.java
package year2019;

import template.Day;

public class Day04 extends Day {

    private final int input = Integer.parseInt(getInput().get(0).substring(0, 6));
    private final int input2 = Integer.parseInt(getInput().get(0).substring(getInput().get(0).length() - 6));

    public Day04() {
        super("input.txt", "04", "2019");
    }
    private boolean validPassword(int password){
        String passwordStr = String.valueOf(password);
        boolean valid = false;
        for (int i = 1; i < passwordStr.length(); i++) {
            if(passwordStr.charAt(i - 1) > passwordStr.charAt(i)){
                return false;
            }
            if(passwordStr.charAt(i - 1) == passwordStr.charAt(i)){
                valid = true;
            }
        }
        return valid;
    }

    private boolean validPassword2(int password) {
        boolean valid = false;
        int count = 1;
        String passwordStr = String.valueOf(password);
        for (int i = 1; i < passwordStr.length(); i++) {
            if (passwordStr.charAt(i - 1) > passwordStr.charAt(i)) {
                return false;
            }
            if (passwordStr.charAt(i - 1) == passwordStr.charAt(i)) {
                count++;
            } else {
                if (count == 2) {
                    valid = true;
                }
                count = 1;
            }
        }
        if (count == 2) {
            valid = true;
        }
        return valid;
    }

    @Override
    public long part1() {
        int count = 0;
        for (int i = input; i < input2; i++) {
            if(validPassword(i)){
                count++;
            }
        }
        return count;
    }

    @Override
    public long part2() {
        int count = 0;
        for (int i = input; i < input2; i++) {
            if(validPassword2(i)){
                count++;
            }
        }
        return count;
    }
}

File: src/main/year2019/day04/Day04.kt
package year2019.day04

import java.io.File

class Day04(text : String = "input.txt") {

    private val day = "04"
    private val year = "2019"

    private val input = File("src/resources/$year/$day/$text").readText().take(6).toInt()
    private val input2 = File("src/resources/$year/$day/$text").readText().takeLast(6).toInt()

    fun part1(): Int = (input..input2).count { c -> (c.toString().zipWithNext().all { it.first <= it.second } && c.toString().zipWithNext().any { it.first == it.second })}
    fun part2(): Int = (input..input2).count { c -> (c.toString().zipWithNext().all { it.first <= it.second } && c.toString().zipWithNext().any { it.first == it.second }) && c.toString().groupingBy { it }.eachCount().any { it.value == 2 } }

}

